DAT-9663 Liquibase Snapshot/Diff fails on Postgres with datatype "real" with default value "DEFAULT (0)::real"

https://datical.atlassian.net/browse/DAT-9663

### Environment

PostgreSQL

## Description

The following Postgresql column data type with the DEFAULT value generate issues with the snapshot/diff/diffchangeLog/generateChangeLog commands.

myColumn real DEFAULT (0)::real

## Steps to reproduce

    Create the following table in the target database:

```
CREATE TABLE community (
  id text NOT NULL DEFAULT,
  "name" text NULL,
  "number" text NULL,
  division text NULL,
  latitude real DEFAULT (0)::real,
  longitude real DEFAULT (0)::real,
  latilong point NULL DEFAULT '(0,0)'::point,
  created_at timestamptz NULL DEFAULT now(),
  updated_at timestamptz NULL
);
```

2. Run the snapshot command against the new change/database.

## Actual results:

Snapshot fails with the following output:

```
liquibase snapshot --log-level=info


Starting Liquibase at 09:14:17 (version 4.8.0 #1581 built at 2022-02-18 21:43+0000)
Liquibase Version: 4.8.0
Liquibase Pro 4.8.0 by Liquibase licensed to customersuccess until Sun Jun 26 23:59:59 CDT 2022



[2022-03-08 09:14:22] SEVERE [liquibase.integration] java.lang.NumberFormatException
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
        at liquibase.command.CommandScope.execute(CommandScope.java:163)
        at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:51)
        at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:21)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at liquibase.integration.commandline.LiquibaseCommandLine.lambda$execute$1(LiquibaseCommandLine.java:325)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:165)
        at liquibase.integration.commandline.LiquibaseCommandLine.execute(LiquibaseCommandLine.java:291)
        at liquibase.integration.commandline.LiquibaseCommandLine.main(LiquibaseCommandLine.java:80)
Caused by: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
        at liquibase.integration.commandline.Main$1.run(Main.java:443)
        at liquibase.integration.commandline.Main$1.run(Main.java:221)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:165)
        at liquibase.integration.commandline.Main.run(Main.java:221)
        at liquibase.command.AbstractCliWrapperCommandStep.run(AbstractCliWrapperCommandStep.java:32)
        at liquibase.command.CommandScope.execute(CommandScope.java:157)
        ... 14 more
Caused by: liquibase.exception.CommandExecutionException: java.lang.NumberFormatException
        at liquibase.command.CommandScope.execute(CommandScope.java:163)
        at liquibase.integration.commandline.Main.doMigration(Main.java:1646)
        at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:398)
        at liquibase.Scope.lambda$child$0(Scope.java:180)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:179)
        at liquibase.Scope.child(Scope.java:158)
        at liquibase.integration.commandline.Main$1.run(Main.java:397)
        ... 20 more
Caused by: java.lang.NumberFormatException
        at java.math.BigDecimal.<init>(BigDecimal.java:497)
        at java.math.BigDecimal.<init>(BigDecimal.java:383)
        at java.math.BigDecimal.<init>(BigDecimal.java:809)
        at liquibase.util.SqlUtil.parseValue(SqlUtil.java:260)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.readDefaultValue(ColumnSnapshotGenerator.java:686)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.readColumn(ColumnSnapshotGenerator.java:394)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.addTo(ColumnSnapshotGenerator.java:204)
        at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:79)
        at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:312)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:445)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:467)
        at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:373)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:335)
        at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:105)
        at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:58)
        at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:34)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:215)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:188)
        at liquibase.command.core.InternalSnapshotCommandStep.run(InternalSnapshotCommandStep.java:97)
        at liquibase.command.CommandScope.execute(CommandScope.java:157)
        ... 27 more

Unexpected error running Liquibase: java.lang.NumberFormatException

liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
        at liquibase.command.CommandScope.execute(CommandScope.java:163)
        at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:51)
        at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:21)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at liquibase.integration.commandline.LiquibaseCommandLine.lambda$execute$1(LiquibaseCommandLine.java:325)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:165)
        at liquibase.integration.commandline.LiquibaseCommandLine.execute(LiquibaseCommandLine.java:291)
        at liquibase.integration.commandline.LiquibaseCommandLine.main(LiquibaseCommandLine.java:80)
Caused by: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
        at liquibase.integration.commandline.Main$1.run(Main.java:443)
        at liquibase.integration.commandline.Main$1.run(Main.java:221)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:165)
        at liquibase.integration.commandline.Main.run(Main.java:221)
        at liquibase.command.AbstractCliWrapperCommandStep.run(AbstractCliWrapperCommandStep.java:32)
        at liquibase.command.CommandScope.execute(CommandScope.java:157)
        ... 14 more
Caused by: liquibase.exception.CommandExecutionException: java.lang.NumberFormatException
        at liquibase.command.CommandScope.execute(CommandScope.java:163)
        at liquibase.integration.commandline.Main.doMigration(Main.java:1646)
        at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:398)
        at liquibase.Scope.lambda$child$0(Scope.java:180)
        at liquibase.Scope.child(Scope.java:189)
        at liquibase.Scope.child(Scope.java:179)
        at liquibase.Scope.child(Scope.java:158)
        at liquibase.integration.commandline.Main$1.run(Main.java:397)
        ... 20 more
Caused by: java.lang.NumberFormatException
        at java.math.BigDecimal.<init>(BigDecimal.java:497)
        at java.math.BigDecimal.<init>(BigDecimal.java:383)
        at java.math.BigDecimal.<init>(BigDecimal.java:809)
        at liquibase.util.SqlUtil.parseValue(SqlUtil.java:260)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.readDefaultValue(ColumnSnapshotGenerator.java:686)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.readColumn(ColumnSnapshotGenerator.java:394)
        at liquibase.snapshot.jvm.ColumnSnapshotGenerator.addTo(ColumnSnapshotGenerator.java:204)
        at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:79)
        at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:312)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:445)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:467)
        at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:373)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:335)
        at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:105)
        at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:58)
        at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:34)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:215)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:188)
        at liquibase.command.core.InternalSnapshotCommandStep.run(InternalSnapshotCommandStep.java:97)
        at liquibase.command.CommandScope.execute(CommandScope.java:157)
        ... 27 more
```
 
### Expected Result

snapshot/diff/diffchangeLog/generateChangeLog commands do not fail when myColumn real DEFAULT (0)::real

--- a/liquibase-core/src/main/java/liquibase/snapshot/jvm/ColumnSnapshotGenerator.java
+++ b/liquibase-core/src/main/java/liquibase/snapshot/jvm/ColumnSnapshotGenerator.java
@@ -37,7 +37,7 @@ public class ColumnSnapshotGenerator extends JdbcSnapshotGenerator {
     protected static final String COLUMN_DEF_COL = "COLUMN_DEF";
 
     private Pattern postgresStringValuePattern = Pattern.compile("'(.*)'::[\\w .]+");
-    private Pattern postgresNumberValuePattern = Pattern.compile("(\\d*)::[\\w .]+");
+    private Pattern postgresNumberValuePattern = Pattern.compile("\\(?(\\d*)\\)?::[\\w .]+");
 
 
     public ColumnSnapshotGenerator() {

--- a/liquibase-core/src/test/groovy/liquibase/snapshot/jvm/ColumnSnapshotGeneratorTest.groovy
+++ b/liquibase-core/src/test/groovy/liquibase/snapshot/jvm/ColumnSnapshotGeneratorTest.groovy
@@ -1,9 +1,13 @@
 package liquibase.snapshot.jvm
 
+import liquibase.database.core.MSSQLDatabase
 import liquibase.database.core.PostgresDatabase
 import liquibase.snapshot.CachedRow
+import liquibase.statement.DatabaseFunction
 import liquibase.structure.core.Column
+import liquibase.structure.core.DataType
 import spock.lang.Specification
+import spock.lang.Unroll
 
 class ColumnSnapshotGeneratorTest extends Specification {
     private ColumnSnapshotGenerator columnSnapshotGenerator
@@ -43,5 +47,21 @@ class ColumnSnapshotGeneratorTest extends Specification {
         then:
         dataType.getColumnSize() == 100
         dataType.getTypeName() == "varchar"
+    }
+
+    @Unroll
+    def "readDefaultValue"() {
+        expect:
+        columnSnapshotGenerator.readDefaultValue(new CachedRow(["COLUMN_DEF": columnValue]), new Column("col").setType(new DataType(datatype)), db) == expected
+
+        where:
+        columnValue          | datatype  | db                     | expected
+        null                 | "varchar" | new MSSQLDatabase()    | null
+        "(NULL)"             | "varchar" | new MSSQLDatabase()    | new DatabaseFunction("null")
+        "3"                  | "int"     | new PostgresDatabase() | 3
+        3                    | "int"     | new PostgresDatabase() | 3
+        "(3)::real"          | "float"   | new PostgresDatabase() | 3
+        "3::real"            | "float"   | new PostgresDatabase() | 3
+        "'a value'::varchar" | "varchar" | new PostgresDatabase() | "a value"
     }
 }

